design/ceph/object/ceph-bucket-notification-crd.md

```diff
@@ -0,0 +1,82 @@
+
+## Overview
+
+Ceph added support for the bucket notifications feature from Nautilus onwards. It allows to send messages to various endpoints when a new event occurs on a bucket (ref: <https://docs.ceph.com/docs/master/radosgw/notifications/)>
+
+Setup of those notifications are normally done by sending HTTP requests to the RGW, either to create/delete topics pointing to specific endpoints, or create/delete bucket notifications based on those topics.
+
+This functionality eases this process by avoiding to use external tools or scripts. It is replaced by creation of CR definitions that contain all the information necessary to create topics and/or notifications, which the rook operator processes.
+
+## Goals
+
+Creates a CRD for topics and a CRD for notifications, defining all the necessary and optional information for the various endpoints.
+
+Extends the rook operator to handle the CRs that would be submitted by users.
+
+## Implementation
+
+The CR for a topic configuration takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephBucketTopic
+metadata:
+  name: # name of the topic
+  namespace: # namespace where topic belongs
+Spec:
+  endpoint: #(mandatory) URI of an endpoint to send push notification to
+  opaqueData: #(optional) opaque data is set in the topic configuration
+  persistent: #(optional) indication whether notifications to this endpoint are persistent or not (`false` by default)
+  http:
+    verifySSL: #indicate whether the server certificate is validated by the client or not (`true` by default)
+  amqp:
+    ackLevel: # none/routable/broker, optional (default - broker)
+    amqpExchange: # exchanges must exist and be able to route messages based on topics
+  kafka:
+    useSSL: # secure connection will be used for connecting with the broker (`false` by default)
+    caLocation: # this specified CA will be used, instead of the default one, to authenticate the broker
+    ackLevel: # none/broker, optional (default - broker)
+```
+
+The CR for bucket notification takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: cephBucketNotification
+metadata:
+  name: # name of the notification
+  namespace: # namespace where notification belongs
+Spec:
+  topic: #(mandatory) reference to the topic, topic_arn
+  filter: #(optional) Prefix/Suffix/Regex/Metadata/Tags, optional (default - {})
+    metadata:
+    - name: x-amz-meta-color
+      value: blue
+    stringMatch:
+    - name: prefix
+      value: hello
+    - name: suffix
+      value: .png
+    - name: regex
+      value: [a-z]*
+
+  Events: # vaules appilicable [here](https://docs.ceph.com/en/latest/radosgw/s3-notification-compatibility/#event-types), (default all)
+```
+
+After creating resources for notifications, it can be referred in `Additional Config` field of OBC. `Additional Config` is string map of key value pairs and notification can be referred as `bucketNotification` key. It will be list of notification separated by comma.
```

Comment:
The bucket notification is defined with the CRD, then it is applied to a bucket in the additionalConfig of the OBC, right? So you could have the same notification applied to different buckets? Sounds good.

Reply:
Yes correct that's intention :)